storage: harden replica write backpressure mechanism

nvb opened the pull request:
Fixes #24215.
Fixes #25036.
Fixes #24974.
Fixes #25141.
Fixes #25142.

This PR includes a change to harden the replica's write backpressure
mechanism. We now reject backpressured write when a range is unsplittable.
This is important for placing an upper bound on the size of a range when it
consists of just a single row (#24215).

The change also fixes a few flaky tests that resulted from backpressure
not being restrictive enough.

nvb committed:
storage: reject backpressured write when range is unsplittable

Fixes #24215.

This change rejects requests to unsplittable ranges that need to be
backpressured immediately. It does so by looking at the split queue
purgatory and rejecting backpressured writes to any range that is
in purgatory.

Release note: None

nvb committed:
roachtest: un-skip hotspots splits test

Fixes #25036.

This fixes the flakes in the test. They have passed 5 straight times for
me. It's also easy to prove that backpressure is necessary for the test
to pass, just run the following and it quickly fails:

```
set cluster setting kv.range.backpressure_range_size_multiplier=0;
```

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/25261)
<!-- Reviewable:end -->

nvb committed:
storage: fix TestStoreRangeSplitBackpressureWrites

Fixes #24974.
Fixes #25141.
Fixes #25142.

This change fixes TestStoreRangeSplitBackpressureWrites and
increases the its testing area to include cases where a range
is large enough to require backpressure but does not have a
split in-flight.

Release note: None

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 8 files at r7, 2 of 2 files at r8, 1 of 1 files at r9.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/25261#-:-LCesbCn2epsjNhE-Dd_:bnfp4nl)*
<!-- Sent from Reviewable.io -->
